### PR TITLE
fix: 15131 html decode in lhn and iou

### PR DESCRIPTION
**Pull request by tienifr**
### Fixed Issues

$ https://github.com/Expensify/App/issues/16819
$ https://github.com/Expensify/App/issues/15131
PROPOSAL: https://github.com/Expensify/App/issues/15131#issuecomment-1497019078

### Tests

- LHN should show the last message
1. Go to chat
2. Enter `&nbsp`
3. Verify LHN should show `&nbsp`
4. Enter `*abc*`
5. Verify LHN should show `*abc*`
- IOU cancelling message shouldn't display escaped html characters
1. Go to chat
2. Request money with IOU message contains HTML characters, like Tokyo <> Mexico
3. Click sent IOU message to open IOU details page
4. Click cancel money request
5. Verify IOU cancelling message should display Tokyo <> Mexico

- iou message should not be flicker when user goes offline then online again
1. Go to chat
2. Go offline
3. Request money
4. Enter amount and click Next
5. Type What's it for message: test *abc* &nbsp
6. Click request button
7. Verify the message in report should be test abc &nbsp
8. Go online
9. Verify the message in report should be test abc &nbsp and shouldn't be flicker

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as above

### QA Steps

- LHN should show the last message
1. Go to chat
7. Enter `&nbsp`
8. Verify LHN should show `&nbsp`
13. Enter `*abc*`
14. Verify LHN should show `*abc*`
- IOU cancelling message shouldn't display escaped html characters
1. Go to chat
2. Request money with IOU message contains HTML characters, like Tokyo <> Mexico
3. Click sent IOU message to open IOU details page
4. Click cancel money request
5. Verify IOU cancelling message should display Tokyo <> Mexico

- iou message should not be flicker when user goes offline then online again
1. Go to chat
2. Go offline
3. Request money
4. Enter amount and click Next
5. Type What's it for message: test *abc* &nbsp
6. Click request button
7. Verify the message in report should be test abc &nbsp
8. Go online
9. Verify the message in report should be test abc &nbsp and shouldn't be flicker
- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/113963320/230038158-7c1789d2-87db-4773-b45b-1e092853cbe5.mov

iou message should not be flicker when user goes offline then online again:

https://user-images.githubusercontent.com/113963320/231097630-d89223a0-7996-4956-9e06-1c31c9723942.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/113963320/230040355-67afaad4-be36-4b74-b855-caaae53d910c.mp4
iou message should not be flicker when user goes offline then online again:

https://user-images.githubusercontent.com/113963320/231099912-5344759f-9d50-4216-9f7e-9a6f100ffc02.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/113963320/230039178-f9eb243b-4c29-47f8-82bc-77bc21ce099a.mp4

iou message should not be flicker when user goes offline then online again:

https://user-images.githubusercontent.com/113963320/231099748-8dbb2ba3-e1b6-469f-9c38-cacc33381286.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/113963320/230041357-a7ceb2a9-aa3c-4b43-9c20-6c06a71d98da.mp4
iou message should not be flicker when user goes offline then online again:

https://user-images.githubusercontent.com/113963320/231102273-c1e46d3e-d8e6-403c-a7bb-b173a76708dc.mp4

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/113963320/231127138-cf194904-0683-45f3-a064-987e645bb5ec.mp4

iou message should not be flicker when user goes offline then online again:

https://user-images.githubusercontent.com/113963320/231127122-e87d0aeb-7f66-4cfc-b4b6-b2d8919847cf.mp4

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/113963320/231190586-6e1c36a8-abfb-4746-b2cc-432464178817.mp4

https://user-images.githubusercontent.com/113963320/231190605-9d9a0ce6-a7ae-4449-a63c-b94def6c55fa.mp4

</details>

### CODE REVIEW

**Review by sobitneupane — COMMENTED** [review]
Changes look good. But let's make sure it works in all test conditions.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/25876548/230075078-7b2909af-c4dd-47b5-ba20-ce39a9631ec0.mov

</details>


<details>
<summary>Desktop</summary>
Screenshot 2023-04-05 at 17 54 06
Screenshot 2023-04-05 at 17 53 35


</details>

**Comment by sobitneupane** [comment]
@tienifr 
Found a bug.

Being offline, If you request money first and the cancel money, html entities (like &nbsp, &#38) will de decoded.

**Comment by rushatgabhane** [comment]
@sobitneupane all yours to review :)

**Comment by tienifr** [comment]
@sobitneupane After investigating I see the problem in https://github.com/Expensify/App/blob/64f4a7814e87bd7c4d085776abf5cb8a67b9e8bc/src/libs/actions/IOU.js#L556
If user is offline, the originalMessage will be `&nbsp`

Screen Shot 2023-04-07 at 17 45 34

but after go online, it's updated to `&amp;nbsp`
Screen Shot 2023-04-07 at 17 46 48

That makes the inconsistency between online and offline mode.

Solution: We can update function `buildOptimisticIOUReportAction` by changing the originalMessage: 
 ```
    const parser = new ExpensiMark();
    const commentText = getParsedComment(comment);
    const textForNewComment = parser.htmlToText(commentText)
    const textFotNewCommentDecode = Str.htmlDecode(textForNewComment)
    const originalMessage = {
        amount,
        comment:textForNewComment,
        currency,
        IOUTransactionID,
        IOUReportID,
        type,
    };

...
        message: getIOUReportActionMessage(type, amount, participants, textFotNewCommentDecode, currency, paymentType, isSettlingUp),

```
But I don't think it's a good ideal, we should preserve the comment as its name `originalMessage`.
Please tell me what your thought. Thanks

**Comment by sobitneupane** [comment]
@tienifr I think we should change the frontend to match the backend if it won't cause any regression.

**Comment by tienifr** [comment]
@sobitneupane I updated my PR to fix this issue: https://github.com/Expensify/App/pull/16957#issuecomment-1498563710
Here is the result:

https://user-images.githubusercontent.com/113963320/230857440-dcef6a9d-914f-43b6-828e-ae2bb08d8c22.mp4

**Comment by sobitneupane** [comment]
Thanks for the update @tienifr. Can you please update the screenshots.

**Comment by tienifr** [comment]
@sobitneupane I've added the screenshots, and PR body. Please help take a look. Thanks

**Review by sobitneupane — COMMENTED** [review]
### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/25876548/231391253-a7f1eaa9-101f-4edd-9229-5760c7fcf2d4.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/25876548/231396889-c9717350-2208-45a6-abf1-5882a20b6d5f.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/25876548/231396303-b27f1c51-82e4-40bb-ae85-5ddf9da3dacd.mov


</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/25876548/231393032-994a5ceb-80b4-49d1-a7f4-7131e847c274.mov

</details>

<details>
<summary>iOS</summary>


https://user-images.githubusercontent.com/25876548/231396334-ea45aab7-dccd-4a35-8640-146fa32c751d.mov


</details>

<details>
<summary>Android</summary>


https://user-images.githubusercontent.com/25876548/231396349-42070c88-92a4-45b9-bda2-06d5e0249e0b.mov


</details>

**Review by sobitneupane — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

**Review by MariaHCD — APPROVED** [review]

**Review by pecanoro — COMMENTED** [review]
@tienifr So the PR works well, but I am wondering why we decided to do all this dance with htmlDecode and we are not repeating what we do with requestMoney for example? https://github.com/Expensify/App/blob/604c759196e520fe93003d25d2c7ed092d5ae174/src/libs/actions/IOU.js#L55-L89

**Comment by tienifr** [comment]
@pecanoro With requestMoney we're passing the comment state instead of originalMessage: https://github.com/Expensify/App/blob/42bea4bd7bd18231f7b5fb5dfd94ca5b58fe1c2b/src/pages/iou/MoneyRequestModal.js#L312-L319

The originalMessage is encode but not comment state. That why we don't need to decode the comment in this case

**Comment by pecanoro** [comment]
@tienifr Gotcha, thank you for the explanation! Hopefully we solve everything with this and we can forget about it hehe

**Review by pecanoro — APPROVED** [review]

File: src/libs/SidebarUtils.js
Comment (jayeshmangwani) [inline src/libs/SidebarUtils.js:1]: This line causes a crash on iOS, when we get `lastMessageText: null` then it is causing the crash

Simulator Screen Shot - iPhone 14 - 2023-04-14 at 09 44 56

cc: @tienifr @sobitneupane @pecanoro @MariaHCD

File: src/libs/SidebarUtils.js [hunk 1]
```diff
@@ -247,7 +247,7 @@ function getOptionData(reportID) {
     if (ReportUtils.isReportMessageAttachment({text: report.lastMessageText, html: report.lastMessageHtml})) {
         lastMessageTextFromReport = `[${Localize.translateLocal('common.attachment')}]`;
     } else {
-        lastMessageTextFromReport = Str.htmlDecode(report ? report.lastMessageText : '');
+        lastMessageTextFromReport = _.get(report, 'lastMessageText', '');
```
Comment (jayeshmangwani) [inline src/libs/SidebarUtils.js:250]: This line causes a crash on iOS, when we get `lastMessageText: null` then it is causing the crash

Simulator Screen Shot - iPhone 14 - 2023-04-14 at 09 44 56

cc: @tienifr @sobitneupane @pecanoro @MariaHCD

Comment (sobitneupane) [reply src/libs/SidebarUtils.js:250]: @tienifr Let's get this sorted asap.

Comment (sobitneupane) [reply src/libs/SidebarUtils.js:250]: Created PR https://github.com/Expensify/App/pull/17449

Comment (tienifr) [reply src/libs/SidebarUtils.js:250]: @sobitneupane Thanks for your quick PR. I'm really sorry for this delay. Do you need any help?

Comment (sobitneupane) [reply src/libs/SidebarUtils.js:250]: @tienifr If you are available, can you please review the [PR](https://github.com/Expensify/App/pull/17449)

Comment (tienifr) [reply src/libs/SidebarUtils.js:250]: Yes, sure

**Pull request merged**